contrib/mesos/pkg/scheduler/podtask/pod_task.go

```diff
@@ -170,6 +176,25 @@ func (t *T) Has(f FlagType) (exists bool) {
 	return
 }
 
+func (t *T) Roles() []string {
+	if len(t.Pod.ObjectMeta.Labels) == 0 {
+		return defaultRoles
+	}
+
+	roles := strings.Split(t.Pod.ObjectMeta.Labels[annotation.RolesKey], ",")
+	for i, r := range roles {
+		roles[i] = strings.TrimSpace(r)
+	}
+
+	roles = filterRoles(roles, not(emptyRole), not(duplicateRole()))
```

Comment:
Don't like the duplicateRole predicate. Building up the map is O(n*log(n)). Evaluating the filter is O(n^2 * log(n)). Of course we only have a handful of roles. But still this waste is bad style.